prover/proof_submitter/transaction/sender.go

```diff
@@ -0,0 +1,194 @@
+package transaction
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/taikoxyz/taiko-client/bindings/encoding"
+	"github.com/taikoxyz/taiko-client/pkg/rpc"
+	proofProducer "github.com/taikoxyz/taiko-client/prover/proof_producer"
+)
+
+var (
+	ErrUnretryable = errors.New("unretryable")
+)
+
+// Sender is responsible for sending proof submission transactions with a backoff policy, if
+// the transaction should not be retried anymore, it will return an `ErrUnretryable` error.
+type Sender struct {
+	rpc                *rpc.Client
+	backOffPolicy      backoff.BackOff
+	maxRetry           *uint64
+	waitReceiptTimeout time.Duration
+}
+
+// NewSender creates a new Sener instance.
+func NewSender(
+	cli *rpc.Client,
+	retryInterval time.Duration,
+	maxRetry *uint64,
+	waitReceiptTimeout time.Duration,
+) *Sender {
+	var backOffPolicy backoff.BackOff = backoff.NewConstantBackOff(retryInterval)
+	if maxRetry != nil {
+		backOffPolicy = backoff.WithMaxRetries(backOffPolicy, *maxRetry)
+	}
+
+	return &Sender{
+		rpc:                cli,
+		backOffPolicy:      backOffPolicy,
+		maxRetry:           maxRetry,
+		waitReceiptTimeout: waitReceiptTimeout,
+	}
+}
+
+// Send sends the given proof to the TaikoL1 smart contract with a backoff policy, if
+// the transaction should not be retried anymore, it will return an `ErrUnretryable` error.
+func (s *Sender) Send(
+	ctx context.Context,
+	proofWithHeader *proofProducer.ProofWithHeader,
+	txAssembler TxBuiler,
+) error {
+	var (
+		isUnretryableError bool
+		nonce              *big.Int
+	)
+
+	if err := backoff.Retry(func() error {
+		if ctx.Err() != nil {
+			return nil
+		}
+
+		// Check if this proof is still needed to be submitted.
+		ok, err := s.validateProof(ctx, proofWithHeader)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return nil
+		}
+
+		// Assemble the taikoL1.proveBlock transaction.
+		tx, err := txAssembler(nonce)
+		if err != nil {
+			err = encoding.TryParsingCustomError(err)
+			if isSubmitProofTxErrorRetryable(err, proofWithHeader.BlockID) {
+				log.Info("Retry sending TaikoL1.proveBlock transaction", "blockID", proofWithHeader.BlockID, "reason", err)
```

Comment:
maybe "reason" should be changed to "error" for logs, easier to grep for all errors and consistent with other logs we have? also perhaps `log.Warn` ?

Reply:
Changed to `log.Warn`, since retry sending this transaction if there is a retryable error is part of the design, and if it hits `maxRetry`, there will be a `log.Error` outside this function.